File: leetcode/270.closest-binary-search-tree-value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int closestValue(TreeNode* root, double target) {
        TreeNode *cur = root;
        int closest = cur->val;
        
        while (cur)
        {
            if (abs(closest - target) > abs(cur->val - target))
            {
                closest = cur->val;
            }
            
            if (cur->val - target > 0)
            {
                cur = cur->left;
            }
            else
            {
                cur = cur->right;
            }
        }
        
        return closest;
    }
};

File: leetcode/270.closest-binary-search-tree-value.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        cur = root
        closest = cur.val
        
        while cur is not None:
            if abs(closest - target) > abs(cur.val - target):
                closest = cur.val
        
            if cur.val - target > 0:
                cur = cur.left
            else:
                cur = cur.right
                
        return closest
        